False positive for global-variable-not-assigned when using with

### Bug description

```python
# File a.py
"""
The "global-variable-not-assigned" check does not take into account that a "with"
statement can very well change the global variable.
For example, when using a global condition variable by multiple threads:
"""

import threading

COND = threading.Condition()


def func1():
    "Function that is called in one thread"
    # pylint: disable=global-statement
    global COND
    with COND:
        COND.notify()
```


### Configuration

_No response_

### Command used

```shell
pylint a.py
```


### Pylint output

```shell
************* Module a
a.py:16:4: W0602: Using global for 'COND' but no assignment is done (global-variable-not-assigned)
```


### Expected behavior

Pylint realizes that "with" may change the global variable and thus counts it like it was an assignment, and thus does not raise the issue in this case.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.0
Python 3.9.7 (default, Sep  3 2021, 12:45:31) 
[Clang 12.0.0 (clang-1200.0.32.29)]
```


### OS / Environment

macOS

### Additional dependencies

Output of pipdeptree:
pylint==1.6.4
  - astroid [required: >=1.4.5,<1.5.0, installed: 1.4.9]
    - lazy-object-proxy [required: Any, installed: 1.4.2]
    - six [required: Any, installed: 1.15.0]
    - wrapt [required: Any, installed: 1.11.2]
  - backports.functools-lru-cache [required: Any, installed: 1.6.4]
  - configparser [required: Any, installed: 4.0.2]
  - isort [required: >=4.2.5, installed: 4.2.15]
  - mccabe [required: Any, installed: 0.6.1]
  - six [required: Any, installed: 1.15.0]

--- a/doc/user_guide/checkers/features.rst
+++ b/doc/user_guide/checkers/features.rst
@@ -1363,9 +1363,9 @@ Variables checker Messages
 :unused-variable (W0612): *Unused variable %r*
   Used when a variable is defined but not used.
 :global-variable-not-assigned (W0602): *Using global for %r but no assignment is done*
-  Used when a name that is previously defined with a "global" statement is
-  being invoked with the `global` keyword but is not reassigned inthe current
-  code block. For example, modifying a global dict.
+  When a variable defined in the global scope is modified in an inner scope, the
+  'global' keyword is required in the inner scope only if there is an
+  assignment operation done in the inner scope.
 :undefined-loop-variable (W0631): *Using possibly undefined loop variable %r*
   Used when a loop variable (i.e. defined by a for loop or a list comprehension
   or a generator expression) is used outside the loop.

--- a/doc/user_guide/checkers/features.rst
+++ b/doc/user_guide/checkers/features.rst
@@ -1363,8 +1363,9 @@ Variables checker Messages
 :unused-variable (W0612): *Unused variable %r*
   Used when a variable is defined but not used.
 :global-variable-not-assigned (W0602): *Using global for %r but no assignment is done*
-  Used when a variable is defined through the "global" statement but no
-  assignment to this variable is done.
+  Used when a name that is previously defined with a "global" statement is
+  being invoked with the `global` keyword but is not reassigned inthe current
+  code block. For example, modifying a global dict.
 :undefined-loop-variable (W0631): *Using possibly undefined loop variable %r*
   Used when a loop variable (i.e. defined by a for loop or a list comprehension
   or a generator expression) is used outside the loop.

--- a/doc/whatsnew/fragments/5073.other
+++ b/doc/whatsnew/fragments/5073.other
@@ -1,0 +1,3 @@
+Update explanation for ``global-variable-not-assigned`` and add inference.
+
+Closes #5073

--- a/doc/whatsnew/fragments/5073.other
+++ b/doc/whatsnew/fragments/5073.other
@@ -1,3 +1,3 @@
-Update explanation for ``global-variable-not-assigned`` and add inference.
+Update explanation for ``global-variable-not-assigned`` and add confidence.
 
 Closes #5073

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -434,8 +434,9 @@ MSGS: dict[str, MessageDefinitionTuple] = {
     "W0602": (
         "Using global for %r but no assignment is done",
         "global-variable-not-assigned",
-        'Used when a variable is defined through the "global" statement '
-        "but no assignment to this variable is done.",
+        'Used when a name that is previously defined with a "global" statement '
+        "is being invoked with the `global` keyword but is not reassigned in"
+        "the current code block. For example, modifying a global dict.",
     ),
     "W0603": (
         "Using the global statement",  # W0121

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -434,9 +434,9 @@ MSGS: dict[str, MessageDefinitionTuple] = {
     "W0602": (
         "Using global for %r but no assignment is done",
         "global-variable-not-assigned",
-        'Used when a name that is previously defined with a "global" statement '
-        "is being invoked with the `global` keyword but is not reassigned in"
-        "the current code block. For example, modifying a global dict.",
+        "When a variable defined in the global scope is modified in an inner scope, "
+        "the 'global' keyword is required in the inner scope only if there is an "
+        "assignment operation done in the inner scope.",
     ),
     "W0603": (
         "Using the global statement",  # W0121

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1480,7 +1480,7 @@ class VariablesChecker(BaseChecker):
         """Check names imported exists in the global scope."""
         frame = node.frame(future=True)
         if isinstance(frame, nodes.Module):
-            self.add_message("global-at-module-level", node=node)
+            self.add_message("global-at-module-level", node=node, confidence=HIGH)
             return
 
         module = frame.root()
@@ -1502,7 +1502,12 @@ class VariablesChecker(BaseChecker):
                 and not utils.is_deleted_after_current(node, name)
                 and not_defined_locally_by_import
             ):
-                self.add_message("global-variable-not-assigned", args=name, node=node)
+                self.add_message(
+                    "global-variable-not-assigned",
+                    args=name,
+                    node=node,
+                    confidence=HIGH,
+                )
                 default_message = False
                 continue
 
@@ -1525,11 +1530,16 @@ class VariablesChecker(BaseChecker):
             else:
                 if not_defined_locally_by_import:
                     # global undefined at the module scope
-                    self.add_message("global-variable-undefined", args=name, node=node)
+                    self.add_message(
+                        "global-variable-undefined",
+                        args=name,
+                        node=node,
+                        confidence=HIGH,
+                    )
                     default_message = False
 
         if default_message:
-            self.add_message("global-statement", node=node)
+            self.add_message("global-statement", node=node, confidence=HIGH)
 
     def visit_assignname(self, node: nodes.AssignName) -> None:
         if isinstance(node.assign_type(), nodes.AugAssign):

--- a/tests/functional/g/globals.py
+++ b/tests/functional/g/globals.py
@@ -4,6 +4,10 @@
 global CSTE  # [global-at-module-level]
 print(CSTE)  # [undefined-variable]
 
+
+RAN_DB_SET = set()
+RAN_DB_DICT = {}
+
 CONSTANT = 1
 def FUNC():
     pass
@@ -90,3 +94,11 @@ def override_class():
         pass
 
     CLASS()
+
+
+def init_connection_state(alias):
+    """Demonstrate that non-assignment modifications to global objects should emit message."""
+    global RAN_DB_SET  # [global-variable-not-assigned]
+    global RAN_DB_DICT  # [global-variable-not-assigned]
+    RAN_DB_SET.add(alias)
+    return RAN_DB_DICT.setdefault("color", "Palomino")

--- a/tests/functional/g/globals.txt
+++ b/tests/functional/g/globals.txt
@@ -1,14 +1,16 @@
 global-at-module-level:4:0:4:11::Using the global statement at the module level:UNDEFINED
 undefined-variable:5:6:5:10::Undefined variable 'CSTE':UNDEFINED
-global-statement:16:4:16:19:fix_contant:Using the global statement:UNDEFINED
-global-variable-not-assigned:23:4:23:14:other:Using global for 'HOP' but no assignment is done:UNDEFINED
-undefined-variable:24:10:24:13:other:Undefined variable 'HOP':UNDEFINED
-global-variable-undefined:29:4:29:18:define_constant:Global variable 'SOMEVAR' undefined at the module level:UNDEFINED
-global-statement:35:4:35:14:global_with_import:Using the global statement:UNDEFINED
-global-statement:41:4:41:21:global_with_import_from:Using the global statement:UNDEFINED
-global-variable-not-assigned:47:4:47:19:global_no_assign:Using global for 'CONSTANT' but no assignment is done:UNDEFINED
-global-statement:53:4:53:19:global_del:Using the global statement:UNDEFINED
-global-statement:60:4:60:19:global_operator_assign:Using the global statement:UNDEFINED
-global-statement:67:4:67:19:global_function_assign:Using the global statement:UNDEFINED
-global-statement:77:4:77:15:override_func:Using the global statement:UNDEFINED
-global-statement:87:4:87:16:override_class:Using the global statement:UNDEFINED
+global-statement:20:4:20:19:fix_contant:Using the global statement:UNDEFINED
+global-variable-not-assigned:27:4:27:14:other:Using global for 'HOP' but no assignment is done:UNDEFINED
+undefined-variable:28:10:28:13:other:Undefined variable 'HOP':UNDEFINED
+global-variable-undefined:33:4:33:18:define_constant:Global variable 'SOMEVAR' undefined at the module level:UNDEFINED
+global-statement:39:4:39:14:global_with_import:Using the global statement:UNDEFINED
+global-statement:45:4:45:21:global_with_import_from:Using the global statement:UNDEFINED
+global-variable-not-assigned:51:4:51:19:global_no_assign:Using global for 'CONSTANT' but no assignment is done:UNDEFINED
+global-statement:57:4:57:19:global_del:Using the global statement:UNDEFINED
+global-statement:64:4:64:19:global_operator_assign:Using the global statement:UNDEFINED
+global-statement:71:4:71:19:global_function_assign:Using the global statement:UNDEFINED
+global-statement:81:4:81:15:override_func:Using the global statement:UNDEFINED
+global-statement:91:4:91:16:override_class:Using the global statement:UNDEFINED
+global-variable-not-assigned:101:4:101:21:init_connection_state:Using global for 'RAN_DB_SET' but no assignment is done:UNDEFINED
+global-variable-not-assigned:102:4:102:22:init_connection_state:Using global for 'RAN_DB_DICT' but no assignment is done:UNDEFINED

--- a/tests/functional/g/globals.txt
+++ b/tests/functional/g/globals.txt
@@ -1,16 +1,16 @@
-global-at-module-level:4:0:4:11::Using the global statement at the module level:UNDEFINED
+global-at-module-level:4:0:4:11::Using the global statement at the module level:HIGH
 undefined-variable:5:6:5:10::Undefined variable 'CSTE':UNDEFINED
-global-statement:20:4:20:19:fix_contant:Using the global statement:UNDEFINED
-global-variable-not-assigned:27:4:27:14:other:Using global for 'HOP' but no assignment is done:UNDEFINED
+global-statement:20:4:20:19:fix_contant:Using the global statement:HIGH
+global-variable-not-assigned:27:4:27:14:other:Using global for 'HOP' but no assignment is done:HIGH
 undefined-variable:28:10:28:13:other:Undefined variable 'HOP':UNDEFINED
-global-variable-undefined:33:4:33:18:define_constant:Global variable 'SOMEVAR' undefined at the module level:UNDEFINED
-global-statement:39:4:39:14:global_with_import:Using the global statement:UNDEFINED
-global-statement:45:4:45:21:global_with_import_from:Using the global statement:UNDEFINED
-global-variable-not-assigned:51:4:51:19:global_no_assign:Using global for 'CONSTANT' but no assignment is done:UNDEFINED
-global-statement:57:4:57:19:global_del:Using the global statement:UNDEFINED
-global-statement:64:4:64:19:global_operator_assign:Using the global statement:UNDEFINED
-global-statement:71:4:71:19:global_function_assign:Using the global statement:UNDEFINED
-global-statement:81:4:81:15:override_func:Using the global statement:UNDEFINED
-global-statement:91:4:91:16:override_class:Using the global statement:UNDEFINED
-global-variable-not-assigned:101:4:101:21:init_connection_state:Using global for 'RAN_DB_SET' but no assignment is done:UNDEFINED
-global-variable-not-assigned:102:4:102:22:init_connection_state:Using global for 'RAN_DB_DICT' but no assignment is done:UNDEFINED
+global-variable-undefined:33:4:33:18:define_constant:Global variable 'SOMEVAR' undefined at the module level:HIGH
+global-statement:39:4:39:14:global_with_import:Using the global statement:HIGH
+global-statement:45:4:45:21:global_with_import_from:Using the global statement:HIGH
+global-variable-not-assigned:51:4:51:19:global_no_assign:Using global for 'CONSTANT' but no assignment is done:HIGH
+global-statement:57:4:57:19:global_del:Using the global statement:HIGH
+global-statement:64:4:64:19:global_operator_assign:Using the global statement:HIGH
+global-statement:71:4:71:19:global_function_assign:Using the global statement:HIGH
+global-statement:81:4:81:15:override_func:Using the global statement:HIGH
+global-statement:91:4:91:16:override_class:Using the global statement:HIGH
+global-variable-not-assigned:101:4:101:21:init_connection_state:Using global for 'RAN_DB_SET' but no assignment is done:HIGH
+global-variable-not-assigned:102:4:102:22:init_connection_state:Using global for 'RAN_DB_DICT' but no assignment is done:HIGH

--- a/tests/functional/r/regression/regression___file___global.txt
+++ b/tests/functional/r/regression/regression___file___global.txt
@@ -1,2 +1,2 @@
-global-statement:5:4:5:19:func:Using the global statement:UNDEFINED
+global-statement:5:4:5:19:func:Using the global statement:HIGH
 redefined-builtin:5:4:5:19:func:Redefining built-in '__file__':UNDEFINED

--- a/tests/functional/u/unused/unused_variable.txt
+++ b/tests/functional/u/unused/unused_variable.txt
@@ -13,7 +13,7 @@ unused-import:55:4:55:38:unused_import_from:Unused namedtuple imported from coll
 unused-import:59:4:59:40:unused_import_in_function:Unused hexdigits imported from string:UNDEFINED
 unused-variable:64:4:64:10:hello:Unused variable 'my_var':UNDEFINED
 unused-variable:75:4:75:8:function:Unused variable 'aaaa':UNDEFINED
-global-statement:97:4:97:39:test_global:Using the global statement:UNDEFINED
+global-statement:97:4:97:39:test_global:Using the global statement:HIGH
 unused-import:103:4:103:28:test_global:Unused platform imported from sys:UNDEFINED
 unused-import:104:4:104:38:test_global:Unused version imported from sys as VERSION:UNDEFINED
 unused-import:105:4:105:15:test_global:Unused import this:UNDEFINED

